src/urbit/lib/chess.hoon

```diff
@@ -399,13 +407,103 @@
+      fen-to-black-castle  ::  black-can-castle
+      fen-to-en-passant    ::  en-passant
+      dem                  ::  ply-50-move-rule
+      dem                  ::  move-number
+  ==
+++  fen-to-board
+  %+  cook  board-helper
+  %+  more  (just '/')
+  ::  XX is this actually slower than alf and fail on invalid char?
+  (stun [1 8] (mask "12345678bknpqrBKNPQR"))
+++  fen-to-player
+  ;~  pose
+      (cold %white (just 'w'))
+      (cold %black (just 'b'))
+  ==
+++  fen-to-white-castle
+  ;~  pose
+      (white-castle-helper hep " -" %none)
+      (white-castle-helper (jest 'k') " k" %none)
+      (white-castle-helper (jest 'q') " q" %none)
+      (white-castle-helper (jest 'KQ') " " %both)
+      (white-castle-helper (just 'K') " " %kingside)
+      (white-castle-helper (just 'Q') " " %queenside)
+  ==
+++  fen-to-black-castle
+  ;~  pose
+      (cold %none hep)
+      (cold %both (jest 'kq'))
+      (cold %kingside (just 'k'))
+      (cold %queenside (just 'q'))
+  ==
+++  fen-to-en-passant
+  ;~  pose
+      (cold ~ hep)
+      %+  cook
+          some
+          ;~  plug
+              (cook term (shim 'a' 'h'))
+              (cook term (mask "36"))
+          ==
+  ==
+++  board-helper
+  |=  brd=(list tape)
+  ?>  =((lent brd) 8)
+  =/  rank=chess-rank  %8
+  =/  file=$?(chess-file %i)  %a
+  =/  board  *(map chess-square chess-piece)
+  |-
+  ?~  brd
+    board
+    ?-  file
+       %i
+     $(brd t.brd, rank (chess-rank (sub rank 1)), file %a)
+   ::
+       chess-file
+     =*  next  (snag (sub file 97) i.brd)
+  ?:  (lte next 56)
+    $(file (chess-file (add file (sub next 48))))
+  ?:  (gte next 97)
+    ?+  (term next)  !!
+      %b  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %bishop]]))
+      %k  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %king]]))
+      %n  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %knight]]))
+      %p  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %pawn]]))
+      %q  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %queen]]))
+      %r  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %rook]]))
+    ==
+  ?+  (term (add next 32))  !!
+    %b  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %bishop]]))
+    %k  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %king]]))
+    %n  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %knight]]))
+    %p  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %pawn]]))
+    %q  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %queen]]))
+    %r  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %rook]]))
+  ==
+==
+++  white-castle-helper
+  |=  [rul=rule pre=tape res=@tas]
+  ^-  rule
+  ;~(pfix rul (funk pre (easy res)))
```

Comment:
This gate is the root cause of the `nest-fail` error - both uses of `rule` are too restrictive, so the compiler accidentally encodes very strict types for what a "rule" is at the point in time when we call it. Code should instead be:
```
|*  [rul=rule pre=tape res=chess-castle]
;~(pfix rul (funk pre (easy res)))
```